projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/asm/LLVMAMD64PushNode.java

```diff
@@ -0,0 +1,93 @@
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.asm;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.CompilerDirectives.CompilationFinal;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.FrameUtil;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.llvm.runtime.memory.LLVMMemory;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack.NeedsStack;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+
+@NeedsStack
+@NodeChild("value")
+public abstract class LLVMAMD64PushNode extends LLVMExpressionNode {
+    @CompilationFinal private FrameSlot stackPointer;
+
+    protected FrameSlot getStackPointerSlot() {
+        if (stackPointer == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            stackPointer = getRootNode().getFrameDescriptor().findFrameSlot(LLVMStack.FRAME_ID);
+            assert stackPointer != null;
+        }
+        return stackPointer;
+    }
+
+    public abstract static class LLVMAMD64PushwNode extends LLVMAMD64PushNode {
+        @Specialization
+        protected Object executeVoid(VirtualFrame frame, short value) {
+            FrameSlot slot = getStackPointerSlot();
+            long sp = FrameUtil.getLongSafe(frame, slot);
+            sp -= 2;
+            frame.setLong(slot, sp);
+            LLVMMemory.putI16(sp, value);
+            return null;
+        }
+    }
+
+    public abstract static class LLVMAMD64PushlNode extends LLVMAMD64PushNode {
+        @Specialization
+        protected Object executeVoid(VirtualFrame frame, int value) {
+            FrameSlot slot = getStackPointerSlot();
+            long sp = FrameUtil.getLongSafe(frame, slot);
+            sp -= 4;
+            frame.setLong(slot, sp);
+            LLVMMemory.putI32(sp, value);
+            return null;
+        }
+    }
+
+    public abstract static class LLVMAMD64PushqNode extends LLVMAMD64PushNode {
+        @Specialization
+        protected Object executeVoid(VirtualFrame frame, long value) {
+            FrameSlot slot = getStackPointerSlot();
+            long sp = FrameUtil.getLongSafe(frame, slot);
+            sp -= 8;
```

Comment:
Do you think that using constants could increase readability for people that are not familiar with x86 assembly? For example, we could introduce a constant `QUAD_BYTE_SIZE` or so.